src/output.rs

```diff
@@ -55,12 +56,23 @@ impl IsPredicate {
 
         if result != self.expected_result {
             if self.expected_result {
-                bail!(ErrorKind::BytesDoesntMatch(
-                    expect.to_owned(),
-                    got.to_owned(),
-                ));
+                Err(AssertionError::new(AssertionKind::OutputMismatch))
+                    .context("expected to match")
+                    .context(KeyValueDisplay::new(
+                        "expect",
+                        DebugDisplay::new(expect.to_owned()),
+                    ))
+                    .context(KeyValueDisplay::new(
+                        "got",
+                        DebugDisplay::new(got.to_owned()),
+                    ))?;
             } else {
-                bail!(ErrorKind::BytesMatches(got.to_owned()));
+                Err(AssertionError::new(AssertionKind::OutputMismatch))
+                    .context("expected to not match")
+                    .context(KeyValueDisplay::new(
+                        "got",
+                        DebugDisplay::new(got.to_owned()),
+                    ))?;
```

Comment:
I used `?` to turn a `Context` into an `Error`.  I felt dirty with it being hard to see in this case.  I somewhat prefer falling back to the old `try!` macro here.

The `throw!` macro that is planned to be added will help with this.

I wonder if `fail` should `impl From<Result<T, Context<D>>> for for Result<T, Error>`. so I could do
```rust
return Err(AssertionKind).into()
```
or maybe `impl From<Fail> for for Result<T, Error>`
```rust
return AssertionKind.into()
```

Reply:
(I think this is also something that would pretty much go away if we used concrete error types)

Reply:
This specific example got cleaned up by the fact that I can now use `bail!`